### PR TITLE
Fix timestep bug

### PR DESCRIPTION
**Pull request by kaelyndunnell**
## Proposed changes

Fixes bug in timestep implementation in accordance with issue #920.

## Types of changes

What types of changes does your code introduce to FESTIM?

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring
- [ ] Documentation Update (if none of the other choices apply)
- [ ] New tests

## Checklist

- [x] Black formatted
- [x] Unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: test/test_stepsize.py [hunk 1]
```diff
@@ -185,7 +185,7 @@ def test_overshoot_milestone(nb_its=1, target=4):
         value=current_value, nb_iterations=nb_its, t=0.9
     )
 
-    expected_value = 1.3
+    expected_value = 1.3 - current_value
```
Comment (RemDelaporteMathurin) [inline test/test_stepsize.py:188]: For this test, you don't even need a growth factor > 1. Try this:

growth factor = 1

curent_dt_value = 100000  very large

current_time = 1  # next timestep will overshoot the milestone of 1.3

expected_value = 1.3 - current_time # you need to fix this !!!

**Review by RemDelaporteMathurin — CHANGES_REQUESTED** [review]
Your test is wrong as you have the wrong expected value

**Review by RemDelaporteMathurin — CHANGES_REQUESTED** [review]
Thanks! CI is passing.

It's failing on the nightly build of FEniCS but we'll wait a bit to see if that gets resolved.

Could you format your code please?

**Comment by jorgensd** [comment]
It’s failing because of the test added by @garth-wells in https://github.com/FEniCS/dolfinx/pull/3502/files 
I am not sure why it shouldn’t be allowed to make a mixed element with a single element.

**Review by RemDelaporteMathurin — APPROVED** [review]

**Comment by garth-wells** [comment]
> It’s failing because of the check added by @garth-wells in https://github.com/FEniCS/dolfinx/pull/3502/files I am not sure why it shouldn’t be allowed to make a mixed element with a single element.

Mixed elements do not have a well-defined value shape, whereas 'base' sub-elements of a mixed element do. An element should be what it is. If it's mixed it should be mixed, if it's not mixed it should not be mixed.

**Comment by jorgensd** [comment]
> > It’s failing because of the check added by @garth-wells in https://github.com/FEniCS/dolfinx/pull/3502/files I am not sure why it shouldn’t be allowed to make a mixed element with a single element.
> 
> Mixed elements do not have a well-defined value shape, whereas 'base' sub-elements of a mixed element do. An element should be what it is. If it's mixed it should be mixed, if it's not mixed it should not be mixed.
@garth-wells 
I still don't see why you shouldn't be allowed to do `mixed_element([single_element])`.
This is a perfectly valid use-case for more generalized problems where one can have `M` number of species in a problem, where 1 species is a special case. It would be silly to have to have dual code paths for handling this in any application code.

**Comment by garth-wells** [comment]
@RemDelaporteMathurin can you share an example of where you rely on a single (base) element to behave like a mixed element?

**Comment by jorgensd** [comment]
> @RemDelaporteMathurin can you share an example of where you rely on a single (base) element to behave like a mixed element?

let’s say you want to support having multiple diffusion equations co-located in space (but with different function spaces. Then you would have a mixed element ([el_0,el_1]). In other parts of the domain (with submesh involved, you might only have a single species. One doesn’t one to treat those differently.

**Comment by RemDelaporteMathurin** [comment]
> @RemDelaporteMathurin can you share an example of where you rely on a single (base) element to behave like a mixed element?

@garth-wells we do not _need_ a single (base) element to behave like a mixed element per se. It would just simplify the logic for us. For example, in the `HydrogenTransportProblem` class, this is how we assign functions (solutions, test functions, previous solutions, etc) to "species":

https://github.com/festim-dev/FESTIM/blob/effcbac86ba0620da1388dcee4e8424ce33939db/src/festim/hydrogen_transport_problem.py#L485-L511

We have this check `if self.multispecies:` in multiple places to check if there are more than one element and if so, treat it as a mixed element.

This is how we do it in the `HTransportProblemDiscontinuous` class (using a mixed domain approach):

https://github.com/festim-dev/FESTIM/blob/effcbac86ba0620da1388dcee4e8424ce33939db/src/festim/hydrogen_transport_problem.py#L990-L1021

Here we can have mixed element with only one element and we don't have to have these checks everywhere.

This is an example for assigning functions to species but we have other places in the code where this happens, for instance assigning BCs:

https://github.com/festim-dev/FESTIM/blob/effcbac86ba0620da1388dcee4e8424ce33939db/src/festim/hydrogen_transport_problem.py#L537-L586

Assigning ICs:

https://github.com/festim-dev/FESTIM/blob/effcbac86ba0620da1388dcee4e8424ce33939db/src/festim/hydrogen_transport_problem.py#L610-L642


Having everything as mixed elements would greatly simplify the logic on our side.

**Pull request merged**